driver-core/src/main/java/com/datastax/driver/core/Clock.java

```diff
@@ -15,25 +15,145 @@
+     * Returns the current time in microseconds.
      *
-     * @return the difference, measured in milliseconds, between the current time and midnight, January 1, 1970 UTC.
-     * @see System#currentTimeMillis()
+     * @return the difference, measured in microseconds, between the current time and and the Epoch
+     * (that is, midnight, January 1, 1970 UTC).
      */
-    long currentTime();
+    long currentTimeMicros();
 }
 
 /**
- * Default implementation of a clock that delegate its calls to the system clock.
+ * Factory that returns the best Clock implementation depending on what native libraries are available in the system.
+ * If LibC is available through JNR, and if the system property {@code com.datastax.driver.USE_NATIVE_CLOCK} is set to {@code true}
+ * (which is the default value), then {@link NativeClock} is returned, otherwise {@link SystemClock} is returned.
+ */
+class ClockFactory {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClockFactory.class);
+
+    private static final String USE_NATIVE_CLOCK_SYSTEM_PROPERTY = "com.datastax.driver.USE_NATIVE_CLOCK";
+
+    static Clock newInstance() {
+        if (Native.isLibCLoaded() && SystemProperties.getBoolean(USE_NATIVE_CLOCK_SYSTEM_PROPERTY, true)) {
+            LOGGER.info("Using native clock to generate timestamps");
+            return new NativeClock();
+        } else {
+            LOGGER.info("Using java.lang.System clock to generate timestamps");
+            return new SystemClock();
+        }
+    }
+
+}
+
+/**
+ * Default implementation of a clock that delegates its calls to the system clock.
+ *
+ * @see System#currentTimeMillis()
  */
 class SystemClock implements Clock {
+
     @Override
-    public long currentTime() {
-        return System.currentTimeMillis();
+    public long currentTimeMicros() {
+        return System.currentTimeMillis() * 1000;
     }
-}
\ No newline at end of file
+
+}
+
+/**
+ * Provides the current time with microseconds precision with some reasonable accuracy through
+ * the use of {@link Native#currentTimeMicros()}.
+ * <p/>
+ * But because calling JNR methods is slightly expensive (roughly 300 ns for {@link Native#currentTimeMicros()}),
+ * we only call it once per {@code UPDATE_INTERVAL_MS}
+ * and add the number of nanoseconds since the last call to get the current time, which is good
+ * enough accuracy for our purpose (see CASSANDRA-6106).
+ * <p/>
+ * This reduces the cost of the call to {@link NativeClock#currentTimeMicros()} to roughly 30 ns, which
+ * is comparable to a call to {@link System#currentTimeMillis()}.
+ */
+class NativeClock implements Clock {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NativeClock.class);
+    private static final long UPDATE_INTERVAL_MS = 1000;
+    private static final TimeFetcher TIME_FETCHER = new TimeFetcher();
+
+    static {
+        Thread updater = new Thread(TIME_FETCHER, "Native Clock Fetcher");
```

Comment:
Question about the use of a thread.  Couldn't we change currentTimeMicros() to check if enough time has elapsed instead?  Or is the overhead of that check too much?